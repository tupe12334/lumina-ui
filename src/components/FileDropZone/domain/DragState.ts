export type DragStateType = 'idle' | 'dragOver' | 'dragLeave'

export class DragState {
  private constructor(private readonly state: DragStateType) {}

  public static create(state: DragStateType = 'idle'): DragState {
    return new DragState(state)
  }

  public getState(): DragStateType {
    return this.state
  }

  public isIdle(): boolean {
    return this.state === 'idle'
  }

  public isDragOver(): boolean {
    return this.state === 'dragOver'
  }

  public isDragLeave(): boolean {
    return this.state === 'dragLeave'
  }

  public isActive(): boolean {
    return this.state === 'dragOver'
  }

  public setDragOver(): DragState {
    return new DragState('dragOver')
  }

  public setDragLeave(): DragState {
    return new DragState('dragLeave')
  }

  public setIdle(): DragState {
    return new DragState('idle')
  }
}