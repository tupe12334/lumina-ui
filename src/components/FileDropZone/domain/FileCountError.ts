import { FileValidationError } from './FileValidationError'

export class FileCountError extends FileValidationError {
  constructor(actualCount: number, maxCount: number) {
    super(`Cannot upload ${actualCount} files. Maximum allowed: ${maxCount}`)
    this.name = 'FileCountError'
  }
}