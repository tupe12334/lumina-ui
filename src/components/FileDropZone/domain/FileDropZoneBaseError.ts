export class FileDropZoneError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly fileId?: string
  ) {
    super(message)
    this.name = 'FileDropZoneError'
  }
}