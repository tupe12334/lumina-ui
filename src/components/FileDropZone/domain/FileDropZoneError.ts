export class FileDropZoneError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly fileId?: string
  ) {
    super(message)
    this.name = 'FileDropZoneError'
  }
}

export class FileValidationError extends FileDropZoneError {
  constructor(message: string, fileId?: string) {
    super(message, 'VALIDATION_ERROR', fileId)
    this.name = 'FileValidationError'
  }
}

export class FileUploadError extends FileDropZoneError {
  constructor(message: string, fileId?: string) {
    super(message, 'UPLOAD_ERROR', fileId)
    this.name = 'FileUploadError'
  }
}

export class FileSizeError extends FileValidationError {
  constructor(actualSize: number, maxSize: number, fileId?: string) {
    const actualSizeMB = Math.round(actualSize / 1024 / 1024)
    const maxSizeMB = Math.round(maxSize / 1024 / 1024)
    super(`File size ${actualSizeMB}MB exceeds maximum allowed size of ${maxSizeMB}MB`, fileId)
    this.name = 'FileSizeError'
  }
}

export class FileTypeError extends FileValidationError {
  constructor(actualType: string, allowedTypes: string[], fileId?: string) {
    const typesList = allowedTypes.join(', ')
    super(`File type "${actualType}" is not allowed. Allowed types: ${typesList}`, fileId)
    this.name = 'FileTypeError'
  }
}

export class FileCountError extends FileValidationError {
  constructor(actualCount: number, maxCount: number) {
    super(`Cannot upload ${actualCount} files. Maximum allowed: ${maxCount}`)
    this.name = 'FileCountError'
  }
}