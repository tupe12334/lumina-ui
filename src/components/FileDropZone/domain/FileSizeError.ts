import { FileValidationError } from './FileValidationError'

export class FileSizeError extends FileValidationError {
  constructor(actualSize: number, maxSize: number, fileId?: string) {
    const actualSizeMB = Math.round(actualSize / 1024 / 1024)
    const maxSizeMB = Math.round(maxSize / 1024 / 1024)
    super(`File size ${actualSizeMB}MB exceeds maximum allowed size of ${maxSizeMB}MB`, fileId)
    this.name = 'FileSizeError'
  }
}