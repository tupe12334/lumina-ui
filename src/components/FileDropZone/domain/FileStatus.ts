import { type FileStatus } from './FileStatusType'

export class FileStatusValue {
  private constructor(private readonly status: FileStatus) {}

  public static create(status?: FileStatus): FileStatusValue {
    const defaultStatus = status || 'pending'
    return new FileStatusValue(defaultStatus)
  }

  public getStatus(): FileStatus {
    return this.status
  }

  public isPending(): boolean {
    return this.status === 'pending'
  }

  public isUploading(): boolean {
    return this.status === 'uploading'
  }

  public isUploaded(): boolean {
    return this.status === 'uploaded'
  }

  public isProcessing(): boolean {
    return this.status === 'processing'
  }

  public isCompleted(): boolean {
    return this.status === 'completed'
  }

  public isFailed(): boolean {
    return this.status === 'failed'
  }

  public isActive(): boolean {
    return this.status === 'uploading' || this.status === 'processing'
  }

  public canRetry(): boolean {
    return this.status === 'failed'
  }
}