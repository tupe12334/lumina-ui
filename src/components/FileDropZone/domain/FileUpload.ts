import { FileStatusValue, type FileStatus } from './FileStatus'

export interface FileUploadData {
  id: string
  file: File
  status: FileStatus
  progress: number
  error?: string
  previewUrl?: string
}

export class FileUpload {
  private constructor(
    private readonly id: string,
    private readonly file: File,
    private readonly status: FileStatusValue,
    private readonly progress: number,
    private readonly error?: string,
    private readonly previewUrl?: string
  ) {}

  public static create(file: File, id?: string): FileUpload {
    const fileId = id || `${file.name}-${file.size}-${Date.now()}`
    const status = FileStatusValue.create('pending')
    return new FileUpload(fileId, file, status, 0)
  }

  public static fromData(data: FileUploadData): FileUpload {
    const status = FileStatusValue.create(data.status)
    return new FileUpload(
      data.id,
      data.file,
      status,
      data.progress,
      data.error,
      data.previewUrl
    )
  }

  public getId(): string {
    return this.id
  }

  public getFile(): File {
    return this.file
  }

  public getStatus(): FileStatusValue {
    return this.status
  }

  public getProgress(): number {
    return this.progress
  }

  public getError(): string | undefined {
    return this.error
  }

  public getPreviewUrl(): string | undefined {
    return this.previewUrl
  }

  public getName(): string {
    return this.file.name
  }

  public getSize(): number {
    return this.file.size
  }

  public getType(): string {
    return this.file.type
  }

  public withStatus(status: FileStatus): FileUpload {
    const newStatus = FileStatusValue.create(status)
    return new FileUpload(
      this.id,
      this.file,
      newStatus,
      this.progress,
      this.error,
      this.previewUrl
    )
  }

  public withProgress(progress: number): FileUpload {
    const clampedProgress = Math.max(0, Math.min(100, progress))
    return new FileUpload(
      this.id,
      this.file,
      this.status,
      clampedProgress,
      this.error,
      this.previewUrl
    )
  }

  public withError(error: string): FileUpload {
    const failedStatus = FileStatusValue.create('failed')
    return new FileUpload(
      this.id,
      this.file,
      failedStatus,
      this.progress,
      error,
      this.previewUrl
    )
  }

  public withPreviewUrl(previewUrl: string): FileUpload {
    return new FileUpload(
      this.id,
      this.file,
      this.status,
      this.progress,
      this.error,
      previewUrl
    )
  }

  public clearError(): FileUpload {
    return new FileUpload(
      this.id,
      this.file,
      this.status,
      this.progress,
      undefined,
      this.previewUrl
    )
  }

  public toData(): FileUploadData {
    return {
      id: this.id,
      file: this.file,
      status: this.status.getStatus(),
      progress: this.progress,
      error: this.error,
      previewUrl: this.previewUrl
    }
  }
}