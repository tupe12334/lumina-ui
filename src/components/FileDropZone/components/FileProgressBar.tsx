import React from 'react'
import { FileUpload } from '../domain/FileUpload'
import { ProgressBase } from '../../ProgressBase'

interface FileProgressBarProps {
  fileUpload: FileUpload
}

export const FileProgressBar: React.FC<FileProgressBarProps> = ({ fileUpload }) => {
  const status = fileUpload.getStatus()

  if (!status.isUploading() && !status.isProcessing()) {
    return null
  }

  const getVariant = () => {
    if (status.isFailed()) return 'error'
    if (status.isCompleted()) return 'success'
    return 'default'
  }

  return (
    <ProgressBase
      value={fileUpload.getProgress()}
      variant={getVariant()}
      size="small"
      aria-label={`File upload progress: ${fileUpload.getProgress()}%`}
    />
  )
}