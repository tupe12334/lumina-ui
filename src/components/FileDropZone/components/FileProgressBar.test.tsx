import { render, screen } from '@testing-library/react'
import { FileProgressBar } from './FileProgressBar'
import { FileUpload } from '../domain/FileUpload'

const mockFile = new File(['test content'], 'test.txt', { type: 'text/plain' })

describe('FileProgressBar', () => {
  test('renders correctly with uploading file', () => {
    const fileUpload = FileUpload.create(mockFile)
      .withStatus('uploading')
      .withProgress(50)

    const { container } = render(<FileProgressBar fileUpload={fileUpload} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('returns null for non-uploading files', () => {
    const fileUpload = FileUpload.create(mockFile).withStatus('pending')

    const { container } = render(<FileProgressBar fileUpload={fileUpload} />)
    expect(container.firstChild).toBeNull()
  })

  test('shows progress bar for uploading files', () => {
    const fileUpload = FileUpload.create(mockFile)
      .withStatus('uploading')
      .withProgress(75)

    render(<FileProgressBar fileUpload={fileUpload} />)

    const progressBar = screen.getByRole('progressbar')
    expect(progressBar).toBeInTheDocument()
    expect(progressBar).toHaveAttribute('aria-valuenow', '75')
    expect(progressBar).toHaveAttribute('aria-valuemin', '0')
    expect(progressBar).toHaveAttribute('aria-valuemax', '100')
  })

  test('shows progress bar for processing files', () => {
    const fileUpload = FileUpload.create(mockFile)
      .withStatus('processing')
      .withProgress(30)

    render(<FileProgressBar fileUpload={fileUpload} />)

    const progressBar = screen.getByRole('progressbar')
    expect(progressBar).toBeInTheDocument()
    expect(progressBar).toHaveAttribute('aria-valuenow', '30')
  })

  test('applies correct CSS classes based on status', () => {
    const completedFileUpload = FileUpload.create(mockFile)
      .withStatus('completed')
      .withProgress(100)

    render(<FileProgressBar fileUpload={completedFileUpload} />)

    const progressFill = document.querySelector('.progressFill')
    expect(progressFill).toHaveClass('completed')
  })

  test('applies failed styling for failed files', () => {
    const failedFileUpload = FileUpload.create(mockFile)
      .withStatus('failed')
      .withProgress(50)

    render(<FileProgressBar fileUpload={failedFileUpload} />)

    const progressFill = document.querySelector('.progressFill')
    expect(progressFill).toHaveClass('failed')
  })

  test('sets correct width style based on progress', () => {
    const fileUpload = FileUpload.create(mockFile)
      .withStatus('uploading')
      .withProgress(60)

    render(<FileProgressBar fileUpload={fileUpload} />)

    const progressFill = document.querySelector('.progressFill')
    expect(progressFill).toHaveStyle('width: 60%')
  })

  test('handles 0% progress', () => {
    const fileUpload = FileUpload.create(mockFile)
      .withStatus('uploading')
      .withProgress(0)

    render(<FileProgressBar fileUpload={fileUpload} />)

    const progressBar = screen.getByRole('progressbar')
    expect(progressBar).toHaveAttribute('aria-valuenow', '0')

    const progressFill = document.querySelector('.progressFill')
    expect(progressFill).toHaveStyle('width: 0%')
  })

  test('handles 100% progress', () => {
    const fileUpload = FileUpload.create(mockFile)
      .withStatus('uploading')
      .withProgress(100)

    render(<FileProgressBar fileUpload={fileUpload} />)

    const progressBar = screen.getByRole('progressbar')
    expect(progressBar).toHaveAttribute('aria-valuenow', '100')

    const progressFill = document.querySelector('.progressFill')
    expect(progressFill).toHaveStyle('width: 100%')
  })
})