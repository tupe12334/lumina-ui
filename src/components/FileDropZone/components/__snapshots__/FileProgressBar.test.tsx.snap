// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileProgressBar > renders correctly with uploading file 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="50"
  class="_progressBar_3af3de"
  role="progressbar"
>
  <div
    class="_progressFill_3af3de"
    style="width: 50%;"
  />
</div>
`;
